File: encode_and_decode_string/encode_decode.go
package encodeanddecodestring

import (
	"strconv"
)

func encode(arr []string) string {
	var str string
	for i := 0; i < len(arr); i++ {
		str += strconv.Itoa(len(arr[i])) + "#" + arr[i]
	}
	return str
}

func decode(str string, res []string) []string {
	if len(str) == 0 {
		return res
	}

	i := 0
	for str[i] != '#' {
		i += 1
	}

	str_length, _ := strconv.Atoi(str[:i])
	start_of_str := i + 1
	end_of_str := str_length + start_of_str
	curr_str := str[start_of_str: end_of_str]
	res = append(res, curr_str)

	return decode(str[end_of_str:], res)

}

File: encode_and_decode_string/encode_decode.js
function encode(arr) {
    let res = ""
    for (i in arr) {
        const char = arr[i]
        res += char.length + "#" + char
    }

    return res
}


function decode(str, res = null) {
    if (str.length < 1) {
        return res
    }

    if(res == null) {
        res = []
    }

    i = 0
    while(str[i] != "#") {
        i++
    }
    const length_of_curr_str = Number(str.slice(0, i))
    const start_of_str = i + 1
    const end_of_str = start_of_str + length_of_curr_str

    res.push(str.slice(start_of_str, end_of_str))

    
    return decode(str.slice(end_of_str), res)
}

console.log(decode("3#ade5#fatai"))